File: python-app/app/discovery.py
"""Finds small Dutch businesses that list a website, via OpenStreetMap.

Two public, keyless services are used:
- Nominatim: geocodes a city name to a bounding box.
- Overpass API: queries OSM for businesses (shops, crafts, small offices,
  local amenities) inside that box that have a `website` tag.

Both are free community services with strict rate limits, so requests are
kept to one at a time and results are capped.
"""
from __future__ import annotations

import time
from dataclasses import dataclass

import requests

from app.config import NOMINATIM_URL, OVERPASS_URL, REQUEST_TIMEOUT_SECONDS, USER_AGENT

# category label -> list of OSM (key, value) tag filters. These are chosen
# to skew towards small/independent local businesses rather than big chains
# (though chain filtering ultimately depends on OSM data quality).
CATEGORIES: dict[str, list[tuple[str, str]]] = {
    "restaurant": [("amenity", "restaurant")],
    "cafe": [("amenity", "cafe")],
    "hairdresser": [("shop", "hairdresser")],
    "beauty_salon": [("shop", "beauty")],
    "bakery": [("shop", "bakery")],
    "butcher": [("shop", "butcher")],
    "florist": [("shop", "florist")],
    "garage_car_repair": [("shop", "car_repair")],
    "plumber": [("craft", "plumber")],
    "electrician": [("craft", "electrician")],
    "carpenter": [("craft", "carpenter")],
    "dentist": [("amenity", "dentist")],
    "physiotherapist": [("healthcare", "physiotherapist")],
    "law_office": [("office", "lawyer")],
    "accountant": [("office", "accountant")],
    "real_estate": [("office", "estate_agent")],
    "architect": [("office", "architect")],
    "gym": [("leisure", "fitness_centre")],
    "clothing_store": [("shop", "clothes")],
    "furniture_store": [("shop", "furniture")],
}


@dataclass
class DiscoveredBusiness:
    name: str
    website: str
    city: str
    category: str
    address: str
    phone: str
    source: str
    source_id: str


def geocode_city(city: str) -> tuple[float, float, float, float]:
    """Return (south, west, north, east) bounding box for a NL place name."""
    resp = requests.get(
        NOMINATIM_URL,
        params={
            "q": f"{city}, Netherlands",
            "format": "json",
            "limit": 1,
            "countrycodes": "nl",
        },
        headers={"User-Agent": USER_AGENT},
        timeout=REQUEST_TIMEOUT_SECONDS,
    )
    resp.raise_for_status()
    results = resp.json()
    if not results:
        raise ValueError(f"Could not geocode city '{city}' in the Netherlands")
    bbox = results[0]["boundingbox"]  # [south, north, west, east] as strings
    south, north, west, east = (float(v) for v in bbox)
    return south, west, north, east


def _build_overpass_query(bbox: tuple[float, float, float, float], tags: list[tuple[str, str]]) -> str:
    south, west, north, east = bbox
    bbox_str = f"{south},{west},{north},{east}"
    clauses = []
    for key, value in tags:
        clauses.append(f'node["{key}"="{value}"]["website"]({bbox_str});')
        clauses.append(f'way["{key}"="{value}"]["website"]({bbox_str});')
        clauses.append(f'node["{key}"="{value}"]["contact:website"]({bbox_str});')
        clauses.append(f'way["{key}"="{value}"]["contact:website"]({bbox_str});')
    body = "\n  ".join(clauses)
    return f"""
[out:json][timeout:60];
(
  {body}
);
out center 100;
""".strip()


def _address_from_tags(tags: dict) -> str:
    parts = [
        tags.get("addr:street", ""),
        tags.get("addr:housenumber", ""),
    ]
    street = " ".join(p for p in parts if p).strip()
    city_part = tags.get("addr:city", "")
    postcode = tags.get("addr:postcode", "")
    line2 = " ".join(p for p in [postcode, city_part] if p).strip()
    return ", ".join(p for p in [street, line2] if p)


def discover_businesses(city: str, category: str, limit: int = 25) -> list[DiscoveredBusiness]:
    if category not in CATEGORIES:
        raise ValueError(f"Unknown category '{category}'")

    bbox = geocode_city(city)
    time.sleep(1)  # be polite to Nominatim before hitting Overpass

    query = _build_overpass_query(bbox, CATEGORIES[category])
    resp = requests.post(
        OVERPASS_URL,
        data={"data": query},
        headers={"User-Agent": USER_AGENT},
        timeout=60,
    )
    resp.raise_for_status()
    elements = resp.json().get("elements", [])

    results: list[DiscoveredBusiness] = []
    seen_websites: set[str] = set()
    for el in elements:
        tags = el.get("tags", {})
        website = (tags.get("website") or tags.get("contact:website") or "").strip()
        name = (tags.get("name") or "").strip()
        if not website or not name:
            continue
        if not website.startswith("http"):
            website = "https://" + website
        normalized = website.rstrip("/").lower()
        if normalized in seen_websites:
            continue
        seen_websites.add(normalized)

        results.append(
            DiscoveredBusiness(
                name=name,
                website=website,
                city=city,
                category=category,
                address=_address_from_tags(tags),
                phone=(tags.get("phone") or tags.get("contact:phone") or ""),
                source="osm",
                source_id=f"{el.get('type')}/{el.get('id')}",
            )
        )
        if len(results) >= limit:
            break

    return results

File: php-app/includes/discovery.php
<?php
declare(strict_types=1);

/**
 * Finds small Dutch businesses via OpenStreetMap.
 * Nominatim geocodes a city name to a bounding box; Overpass API then
 * queries OSM for businesses inside that box.
 *
 * Note: we deliberately do NOT require a `website` tag. Most small
 * businesses in OSM have no website tag at all, and a business with no
 * website is the single best lead there is — so those are kept and
 * flagged instead of filtered away. (Requiring the tag was why searches
 * so often came back empty.)
 */

const NOMINATIM_URL = 'https://nominatim.openstreetmap.org/search';

/** Public Overpass instances, tried in order until one answers. */
function overpass_endpoints(): array
{
    $configured = app_config()['overpass_endpoints'] ?? [];
    if (!empty($configured) && is_array($configured)) {
        return $configured;
    }
    return [
        'https://overpass-api.de/api/interpreter',
        'https://overpass.kumi.systems/api/interpreter',
        'https://overpass.osm.ch/api/interpreter',
    ];
}

function nominatim_url(): string
{
    return app_config()['nominatim_url'] ?? NOMINATIM_URL;
}

/**
 * category slug => ['label' => Dutch label, 'filters' => [Overpass tag selectors]]
 * Multiple selectors are OR-ed together. Selectors are raw Overpass
 * filters so a category can match a whole key (e.g. any ["craft"]).
 */
function categories(): array
{
    return [
        'all' => [
            'label' => 'Alle bedrijven (breedste zoekopdracht)',
            'filters' => [
                '["shop"]',
                '["craft"]',
                '["office"]',
                '["healthcare"]',
                '["amenity"~"^(restaurant|cafe|bar|pub|fast_food|dentist|doctors|veterinary|pharmacy|driving_school|childcare)$"]',
                '["leisure"~"^(fitness_centre|sports_centre|dance)$"]',
                '["tourism"~"^(hotel|guest_house|bed_and_breakfast|apartment)$"]',
            ],
        ],
        'restaurant' => ['label' => 'Restaurant', 'filters' => ['["amenity"="restaurant"]']],
        'cafe' => ['label' => 'Café / lunchroom', 'filters' => ['["amenity"="cafe"]']],
        'bar_pub' => ['label' => 'Bar / café (kroeg)', 'filters' => ['["amenity"="bar"]', '["amenity"="pub"]']],
        'fast_food' => ['label' => 'Snackbar / afhaal', 'filters' => ['["amenity"="fast_food"]']],
        'hotel_bnb' => [
            'label' => 'Hotel / B&B',
            'filters' => ['["tourism"~"^(hotel|guest_house|bed_and_breakfast|apartment)$"]'],
        ],
        'hairdresser' => ['label' => 'Kapper', 'filters' => ['["shop"="hairdresser"]']],
        'beauty_salon' => [
            'label' => 'Schoonheidssalon / nagelstudio',
            'filters' => ['["shop"="beauty"]', '["shop"="nails"]', '["shop"="massage"]'],
        ],
        'bakery' => ['label' => 'Bakkerij', 'filters' => ['["shop"="bakery"]', '["shop"="pastry"]']],
        'butcher' => ['label' => 'Slagerij', 'filters' => ['["shop"="butcher"]']],
        'florist' => ['label' => 'Bloemist', 'filters' => ['["shop"="florist"]']],
        'garage_car_repair' => [
            'label' => 'Autogarage / autobedrijf',
            'filters' => ['["shop"="car_repair"]', '["shop"="car"]', '["shop"="tyres"]'],
        ],
        'bicycle_shop' => ['label' => 'Fietsenmaker', 'filters' => ['["shop"="bicycle"]']],
        'plumber' => ['label' => 'Loodgieter / installateur', 'filters' => ['["craft"="plumber"]', '["craft"="hvac"]']],
        'electrician' => ['label' => 'Elektricien', 'filters' => ['["craft"="electrician"]']],
        'carpenter' => [
            'label' => 'Timmerman / aannemer',
            'filters' => ['["craft"="carpenter"]', '["craft"="builder"]', '["craft"="joiner"]'],
        ],
        'painter' => ['label' => 'Schilder / stukadoor', 'filters' => ['["craft"="painter"]', '["craft"="plasterer"]']],
        'gardener' => ['label' => 'Hovenier', 'filters' => ['["craft"="gardener"]', '["shop"="garden_centre"]']],
        'cleaning' => ['label' => 'Schoonmaakbedrijf', 'filters' => ['["craft"="cleaning"]', '["shop"="laundry"]', '["shop"="dry_cleaning"]']],
        'dentist' => ['label' => 'Tandarts', 'filters' => ['["amenity"="dentist"]', '["healthcare"="dentist"]']],
        'doctor' => ['label' => 'Huisarts', 'filters' => ['["amenity"="doctors"]', '["healthcare"="doctor"]']],
        'physiotherapist' => ['label' => 'Fysiotherapeut', 'filters' => ['["healthcare"="physiotherapist"]']],
        'veterinary' => ['label' => 'Dierenarts', 'filters' => ['["amenity"="veterinary"]']],
        'law_office' => ['label' => 'Advocaat / notaris', 'filters' => ['["office"="lawyer"]', '["office"="notary"]']],
        'accountant' => [
            'label' => 'Accountant / administratiekantoor',
            'filters' => ['["office"="accountant"]', '["office"="tax_advisor"]', '["office"="financial"]'],
        ],
        'real_estate' => ['label' => 'Makelaar', 'filters' => ['["office"="estate_agent"]']],
        'architect' => ['label' => 'Architect', 'filters' => ['["office"="architect"]']],
        'insurance' => ['label' => 'Verzekeringen / financieel advies', 'filters' => ['["office"="insurance"]']],
        'gym' => ['label' => 'Sportschool', 'filters' => ['["leisure"="fitness_centre"]', '["leisure"="sports_centre"]']],
        'clothing_store' => ['label' => 'Kledingwinkel', 'filters' => ['["shop"="clothes"]', '["shop"="shoes"]']],
        'furniture_store' => ['label' => 'Meubelzaak / interieur', 'filters' => ['["shop"="furniture"]', '["shop"="interior_decoration"]']],
        'jewelry_optician' => ['label' => 'Juwelier / opticien', 'filters' => ['["shop"="jewelry"]', '["shop"="optician"]']],
        'pet_shop' => ['label' => 'Dierenwinkel', 'filters' => ['["shop"="pet"]', '["shop"="pet_grooming"]']],
        'driving_school' => ['label' => 'Rijschool', 'filters' => ['["amenity"="driving_school"]']],
        'childcare' => ['label' => 'Kinderopvang', 'filters' => ['["amenity"="childcare"]', '["amenity"="kindergarten"]']],
    ];
}

/** Categories in the shape the frontend dropdown wants. */
function categories_for_ui(): array
{
    $out = [];
    foreach (categories() as $slug => $cat) {
        $out[] = ['value' => $slug, 'label' => $cat['label']];
    }
    return $out;
}

/** Returns [south, west, north, east] bounding box for a NL place name. */
function geocode_city(string $city): array
{
    $params = http_build_query([
        'q' => "{$city}, Netherlands",
        'format' => 'json',
        'limit' => 1,
        'countrycodes' => 'nl',
    ]);
    $res = http_get(nominatim_url() . '?' . $params, 15);
    if (!$res['ok'] || $res['status'] !== 200) {
        throw new RuntimeException(
            'Kon geen verbinding maken met de OpenStreetMap-zoekdienst (Nominatim): '
            . ($res['error'] ?: "HTTP {$res['status']}")
            . '. Controleer of je hosting uitgaande internetverbindingen toestaat.'
        );
    }
    $results = json_decode($res['body'], true);
    if (!is_array($results) || count($results) === 0) {
        throw new RuntimeException(
            "Plaats '{$city}' niet gevonden in Nederland. Controleer de spelling "
            . '(gebruik de officiële plaatsnaam, bijvoorbeeld "Den Bosch" of "s-Hertogenbosch").'
        );
    }
    $bbox = $results[0]['boundingbox']; // [south, north, west, east] as strings
    return [(float) $bbox[0], (float) $bbox[2], (float) $bbox[1], (float) $bbox[3]];
}

function build_overpass_query(array $bbox, array $filters, int $limit): string
{
    [$south, $west, $north, $east] = $bbox;
    $bboxStr = "{$south},{$west},{$north},{$east}";
    $clauses = [];
    foreach ($filters as $filter) {
        // nwr = node/way/relation in one go.
        $clauses[] = "nwr{$filter}({$bboxStr});";
    }
    $body = implode("\n  ", $clauses);
    $outLimit = max(50, $limit * 4); // over-fetch: many results get filtered out below
    return "[out:json][timeout:90];\n(\n  {$body}\n);\nout center {$outLimit};";
}

/** Queries Overpass, trying each mirror until one succeeds. */
function overpass_query(string $query): array
{
    $errors = [];
    foreach (overpass_endpoints() as $endpoint) {
        $ch = curl_init();
        curl_setopt_array($ch, [
            CURLOPT_URL => $endpoint,
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => http_build_query(['data' => $query]),
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => 90,
            CURLOPT_CONNECTTIMEOUT => 10,
            CURLOPT_USERAGENT => crawler_user_agent(),
        ]);
        $body = curl_exec($ch);
        $status = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $curlError = curl_error($ch);
        curl_close($ch);

        if ($body !== false && $status === 200) {
            $data = json_decode($body, true);
            if (is_array($data) && isset($data['elements'])) {
                return $data['elements'];
            }
            $errors[] = "{$endpoint}: ongeldig antwoord";
            continue;
        }
        $errors[] = "{$endpoint}: " . ($curlError ?: "HTTP {$status}");
    }

    throw new RuntimeException(
        'Alle OpenStreetMap-servers (Overpass) gaven een fout. Dit is meestal tijdelijk — '
        . 'probeer het over een paar minuten opnieuw. Details: ' . implode(' | ', $errors)
    );
}

/** Street + house number only; postcode and city are stored separately. */
function address_from_tags(array $tags): string
{
    return trim(($tags['addr:street'] ?? '') . ' ' . ($tags['addr:housenumber'] ?? ''));
}

function normalize_website(string $website): string
{
    $website = trim($website);
    if ($website === '') {
        return '';
    }
    if (!preg_match('#^https?://#i', $website)) {
        $website = 'https://' . ltrim($website, '/');
    }
    return $website;
}

/** Best human-readable description of what the OSM tags say this business is. */
function business_type_from_tags(array $tags): string
{
    foreach (['shop', 'craft', 'office', 'amenity', 'healthcare', 'leisure', 'tourism'] as $key) {
        if (!empty($tags[$key]) && $tags[$key] !== 'yes') {
            return str_replace('_', ' ', $tags[$key]);
        }
    }
    return '';
}

/**
 * Returns a list of business arrays ready for storage.
 * $onlyWithoutWebsite: when true, only businesses that have no website
 * at all are returned (the highest-value leads).
 */
function discover_businesses(string $city, string $category, int $limit = 25, bool $onlyWithoutWebsite = false): array
{
    $cats = categories();
    if (!isset($cats[$category])) {
        throw new InvalidArgumentException("Onbekende categorie '{$category}'");
    }

    $bbox = geocode_city($city);
    usleep(1_000_000); // be polite to Nominatim before hitting Overpass

    $elements = overpass_query(build_overpass_query($bbox, $cats[$category]['filters'], $limit));

    $results = [];
    $seenWebsites = [];
    $seenNames = [];
    foreach ($elements as $el) {
        $tags = $el['tags'] ?? [];
        $name = trim($tags['name'] ?? '');
        if ($name === '') {
            continue; // unnamed POIs are useless as leads
        }

        $website = normalize_website($tags['website'] ?? ($tags['contact:website'] ?? ($tags['url'] ?? '')));
        $hasWebsite = $website !== '';

        if ($onlyWithoutWebsite && $hasWebsite) {
            continue;
        }

        // De-duplicate: by website when there is one, otherwise by name+street.
        if ($hasWebsite) {
            $key = strtolower(rtrim($website, '/'));
            if (isset($seenWebsites[$key])) {
                continue;
            }
            $seenWebsites[$key] = true;
        } else {
            $key = strtolower($name . '|' . ($tags['addr:street'] ?? '') . ($tags['addr:housenumber'] ?? ''));
            if (isset($seenNames[$key])) {
                continue;
            }
            $seenNames[$key] = true;
        }

        $results[] = [
            'name' => $name,
            'website' => $website,
            'has_website' => $hasWebsite,
            'city' => trim($tags['addr:city'] ?? '') !== '' ? $tags['addr:city'] : $city,
            'category' => $category,
            'business_type' => business_type_from_tags($tags),
            'address' => address_from_tags($tags),
            'postcode' => $tags['addr:postcode'] ?? '',
            'phone' => $tags['phone'] ?? ($tags['contact:phone'] ?? ($tags['contact:mobile'] ?? '')),
            'email' => $tags['email'] ?? ($tags['contact:email'] ?? ''),
            'opening_hours' => $tags['opening_hours'] ?? '',
            'facebook' => $tags['contact:facebook'] ?? ($tags['facebook'] ?? ''),
            'instagram' => $tags['contact:instagram'] ?? ($tags['instagram'] ?? ''),
            'lat' => $el['lat'] ?? ($el['center']['lat'] ?? null),
            'lon' => $el['lon'] ?? ($el['center']['lon'] ?? null),
            'source' => 'osm',
            'source_id' => ($el['type'] ?? '') . '/' . ($el['id'] ?? ''),
        ];

        if (count($results) >= $limit) {
            break;
        }
    }

    // Businesses without a website first — they are the strongest leads.
    usort($results, fn($a, $b) => ($a['has_website'] ? 1 : 0) <=> ($b['has_website'] ? 1 : 0));

    return $results;
}
